br/cmd/br/cmd.go

```diff
@@ -162,6 +182,23 @@ func Init(cmd *cobra.Command) (err error) {
 		}
 		log.ReplaceGlobals(lg, p)
 		memory.InitMemoryHook()
+		if debug.SetMemoryLimit(-1) == math.MaxInt64 {
+			memtotal, e := memory.MemTotal()
+			if e != nil {
+				err = e
+				return
+			}
+			memused, e := memory.MemUsed()
+			if e != nil {
+				err = e
+				return
+			}
+			memleft := memtotal - memused
+			memlimit := calculateMemoryLimit(memleft)
+			log.Info("calculate the rest memory",
+				zap.Uint64("memtotal", memtotal), zap.Uint64("memused", memused), zap.Uint64("memlimit", memlimit))
+			debug.SetMemoryLimit(int64(memlimit))
```

Comment:
to be extra safe, can we do a safe cast from uint64 to int64, something like
```
func safeCastUint64ToInt64(u uint64) int64 {
	if u > uint64(math.MaxInt64) {
		return math.MaxInt64
	}
	return int64(u)
}
```
and make it a util method?
cuz in the doc
 `
A zero limit or a limit that's lower than the amount of memory
used by the Go runtime may cause the garbage collector to run
 nearly continuously. However, the application may still make
 progress.
`

Reply:
done. `debug.SetMemoryLimit(int64(memlimit & math.MaxInt64))`